File: go/chapter_3/stack.go
package chapter_3

import "errors"

type stackNode struct {
	value int
	prev  *stackNode
}

type Stack struct {
	top *stackNode
}

func (s *Stack) push(x int) {
	newNode := &stackNode{value: x}
	if s.top != nil {
		newNode.prev = s.top
	}
	s.top = newNode
}

func (s *Stack) pop() (int, error) {
	if s.top == nil {
		return -1, errors.New("Cannot pop. Stack is empty.")
	}
	x := s.top.value
	s.top = s.top.prev
	return x, nil
}

func (s Stack) peek() (int, error) {
	if s.top == nil {
		return -1, errors.New("Cannot peek. Stack is empty.")
	}
	return s.top.value, nil
}

func (s Stack) isEmpty() bool {
	return s.top == nil
}

File: python/chapter_3/stack.py
class StackException(Exception):
    pass


class Stack:
    def __init__(self):
        self.stack = []
        self.length = 0

    def __str__(self):
        return f'stack:{self.stack} length:{self.length}\n'

    def __len__(self):
        return self.length

    def __iter__(self):
        while self.length:
            yield self.pop()

    def push(self, x):
        self.stack.append(x)
        self.length += 1

    def pop(self):
        if self.is_empty():
            raise StackException('Cannot pop. Stack is empty')
        l = self.length - 1
        r = self.stack[l]
        self.stack = self.stack[:l]
        self.length -= 1
        return r

    def peek(self):
        if self.is_empty():
            raise StackException('Cannot peek. Stack is empty')
        return self.stack[self.length - 1]

    def is_empty(self):
        return self.length == 0
